File: default/server/_jscript/Hub.js
(function() {
  var mongodb, node_url, smio;
  var __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; };
  require('./shared/PromiseProxy');
  require('./Database');
  mongodb = require('mongodb');
  node_url = require('url');
  smio = global.smoothio;
  smio.Hub = (function() {
    function Hub(session, url, rc) {
      var inst, server;
      this.session = session;
      this.url = url;
      this.rc = rc;
      this.doc = null;
      server = this.session.server;
      inst = server.inst;
      this.mongo = inst.getDbServer();
      this.dbAdmin = inst.getDb(this.mongo, 'admin');
      this.dbServer = inst.getDb(this.mongo, "smoothio__" + server.serverName);
      this.dbShared = inst.getDb(this.mongo, 'smoothio_shared');
      if (!this.url) {
        this.url = '/';
      }
      this.url = this.url.toLowerCase();
      this.uri = node_url.parse(this.url);
    }
    Hub.prototype.checkExists = function(cb_err_hasHubs) {
      if (this.doc) {
        return cb_err_hasHubs(null, true);
      } else {
        return this.dbServer.withCollection("hubs", __bind(function(err, col) {
          var makeQuery;
          if (err) {
            return cb_err_hasHubs(err);
          }
          makeQuery = function(url) {
            return function() {
              return url.indexOf(this.url) === 0;
            };
          };
          return col.find({
            $where: makeQuery(this.url)
          }).toArray(__bind(function(err, results) {
            if (err) {
              return cb_err_hasHubs(err);
            }
            if (results && results.length) {
              this.doc = _.sortBy(results, function(doc) {
                return -doc.url.length;
              })[0];
              return cb_err_hasHubs(null, true);
            } else {
              return col.find().nextObject(__bind(function(err, doc) {
                return cb_err_hasHubs(err, smio.iif(doc));
              }, this));
            }
          }, this));
        }, this));
      }
    };
    Hub.prototype.getControlUpdates = function(sinceTicks, freq, fresp, cb) {
      var ct;
      if (sinceTicks) {
        return cb(null, {});
      } else {
        ct = "Core_Controls_MainFrame";
        return this.checkExists(function(err, serverHasHubs) {
          if (!serverHasHubs) {
            if (fresp && freq && freq.settings()) {
              fresp.settings({
                bg: '/_/file/images/bg0.jpg'
              });
            }
            ct = "Core_ServerSetup_InitialHubSetup";
          }
          return cb(err, {
            '': {
              '_': ct
            }
          });
        });
      }
    };
    return Hub;
  })();
}).call(this);

File: _cscript/Hub.cs

_ = require 'underscore'
_.mixin require 'underscore.string'
mongodb = require 'mongodb'
node_url = require 'url'

require './shared/PromiseProxy'
require './Database'

smio = global.smoothio

class smio.Hub
	constructor: (@session, @url, @rc) ->
		@doc = null
		server = @session.server
		inst = server.inst
		@mongo = inst.getDbServer()
		@dbAdmin = inst.getDb(@mongo, 'admin')
		@dbServer = inst.getDb(@mongo, "smoothio__#{server.serverName}")
		@dbShared = inst.getDb(@mongo, 'smoothio_shared')
		if not @url
			@url = '/'
		@url = @url.toLowerCase()
		@uri = node_url.parse(@url)

	checkExists: (cb_err_hasHubs) =>
		if @doc
			cb_err_hasHubs(null, true)
		else
			@dbServer.withCollection "hubs", (err, col) =>
				return cb_err_hasHubs(err) if err
				makeQuery = (url) -> -> url.indexOf(this.u) is 0
				col.find($where: makeQuery(@url)).toArray (err, results) =>
					return cb_err_hasHubs(err) if err
					if results and results.length
						@doc = _.sortBy(results, (doc) -> -doc.u.length)[0]
						cb_err_hasHubs(null, true)
					else
						col.find().nextObject (err, doc) =>
							cb_err_hasHubs(err, smio.iif(doc))

	create: (freq, fresp, cb) =>
		@dbServer.withCollection "hubs", (err, col) =>
			return cb(err, col) if err
			col.insert { t: freq.msg.t, u: '/' }, (err, docs) =>
				smio.logit JSON.stringify { e: err, d: docs }
				return cb(err, docs)

	invoke: (cmd, freq, fresp, cb) =>
		if cmd is 'create'
			@create(freq, fresp, cb)
		else
			cb(new Error("Unknown command <b>#{cmd}</b>"))

	getControlUpdates: (sinceTicks, freq, fresp, cb) =>
		return cb(null, '': '_': 'Core_Earth_MainFrame')
		if sinceTicks
			cb(null, {})
		else
			ct = "Core_Controls_MainFrame"
			@checkExists (err, serverHasHubs) ->
				if not serverHasHubs
					if fresp and freq and freq.settings()
						fresp.settings(bg: '/_/file/images/bg0.jpg')
					ct = "Core_ServerSetup_InitialHubSetup"
				cb(err, '': '_': ct)

